analyzer/src/types/ty.rs

```diff
@@ -64,6 +64,7 @@ pub struct FunctionType {
 /// A function parameter.
 #[derive(Clone, Debug, PartialEq)]
 pub struct Parameter {
+    pub(crate) fn_def: Definition,
     pub(crate) segment: Option<SourceSegment>,
```

Comment:
If a `segment` is used, then the `fn_def` is always `Definition::User`.

Define the `Parameter` struct as:
```rs
/// A location in an engine.
#[derive(Clone, Debug, PartialEq)]
pub struct Location {
    pub(crate) definition: SourceId,
    pub(crate) segment: SourceSegment,
}

/// A function parameter.
#[derive(Clone, Debug, PartialEq)]
pub struct Parameter {
    pub(crate) location: Option<Location>,
    pub(crate) ty: TypeId,
}
```

You can then use it at your advantage with the help of pattern matching:
```rs
    if let Some(Location {
        definition: source,
        segment,
    }) = &param.location
    {
        diagnostic = diagnostic.with_observation(
            Observation::new(segment.clone(), *source)
                .with_help("Parameter is declared here")
                .with_tag(ObservationTag::Declaration),
        )
    }
```